### PR TITLE
docs: add metrics-exporter to nfs design doc

### PR DESCRIPTION
**Pull request by synarete**
Signed-off-by: Shachar Sharon <ssharon@redhat.com>

**Description of your changes:**
Extend design doc of ceph-nfs-ganesha to include metrics exporter as sidecar.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: design/ceph/ceph-nfs-ganesha.md [hunk 1]
```diff
@@ -213,6 +213,21 @@ The following directories should **not** be shared between SSSD and other contai
...
+must share the `/run/dbus` directory in order to communicate with the DBus
+daemon via UNIX domain socket within the same pod (similar to way NFS-Ganesha's
+`ganesha_stats` utility communicates with the server).
+
+Exported metrics should include the following details:
+- Exports
+- Connected clients
+- Operations counter per export (for each sub-protocol)
+- I/O stats (write/read/error/other for each sub-protocol)
+- Optional: detailed statistics (for each sub-protocol))
```
Comment (BlaineEXE) [inline design/ceph/ceph-nfs-ganesha.md:229]:
```suggestion
- Optional: detailed statistics (for each sub-protocol)
```

File: design/ceph/ceph-nfs-ganesha.md [hunk 2]
```diff
@@ -213,6 +213,21 @@ The following directories should **not** be shared between SSSD and other contai
...
+monitoring is enable on a k8s cluster, run a sidecar container which will
+provide an end-point for Prometheus metrics scraping. This sidecar container
+must share the `/run/dbus` directory in order to communicate with the DBus
+daemon via UNIX domain socket within the same pod (similar to way NFS-Ganesha's
+`ganesha_stats` utility communicates with the server).
+
+Exported metrics should include the following details:
+- Exports
+- Connected clients
+- Operations counter per export (for each sub-protocol)
```
Comment (BlaineEXE) [inline design/ceph/ceph-nfs-ganesha.md:227]: What is `sub-protocol` in this context? Could we define that here?

**Review by BlaineEXE — COMMENTED** [review]
Overall, I think the basic info is present here that Rook needs.

One thing might be to note what container image the exporter will use. I'm aware that the exporter is currently a WIP and that there might not be a final location. Mentioning that it is a separate image from ceph/ceph might be good, or do we think we want the exporter binary to be present in the ceph/ceph image?

File: design/ceph/ceph-nfs-ganesha.md (hunk 1 above)
Comment (synarete) [reply design/ceph/ceph-nfs-ganesha.md:229]: ACK

File: design/ceph/ceph-nfs-ganesha.md (hunk 2 above)
Comment (synarete) [reply design/ceph/ceph-nfs-ganesha.md:227]: Currently, `NFSv3, NFSv4.0, NFSv4.1, NFSv4.2`. That is, report information at fine-grained NFS protocol granularity.

**Comment by synarete** [comment]
> Overall, I think the basic info is present here that Rook needs.
> 
> One thing might be to note what container image the exporter will use. I'm aware that the exporter is currently a WIP and that there might not be a final location. Mentioning that it is a separate image from ceph/ceph might be good, or do we think we want the exporter binary to be present in the ceph/ceph image?

IMHO, the exporter should be a separate image, running as a sidecar container within the same pod as nfs-ganesha (communicating via dbus socket). Ideally, it should be an upstream code, part of NFS-ganesha project, similar to [ganesha_stats](https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/scripts/ganeshactl/ganesha_stats.py).  It could also be part of ceph or rook, tough probably not the best option.

File: design/ceph/ceph-nfs-ganesha.md [hunk 3]
```diff
@@ -213,6 +213,21 @@ The following directories should **not** be shared between SSSD and other contai
 - `/run/dbus`: using the DBus instance from the sidecar caused SSSD errors in testing. SSSD only
   uses DBus for internal communications and creates its own socket as needed.
 
+### Metrics
+Each NFS-Ganesha server instance exports its stats via DBus. When Prometheus
+monitoring is enable on a k8s cluster, run a sidecar container which will
+provide an end-point for Prometheus metrics scraping. This sidecar container
+must share the `/run/dbus` directory in order to communicate with the DBus
+daemon via UNIX domain socket within the same pod (similar to way NFS-Ganesha's
+`ganesha_stats` utility communicates with the server).
```
[the remark this answers is not in the document]
Comment (travisn) [reply design/ceph/ceph-nfs-ganesha.md:340]: @synarete Link?

**Review by travisn — COMMENTED** [review]
LGTM, just a small suggestion

**Review by travisn — APPROVED** [review]

**Pull request closed without merging**